### PR TITLE
Cleanup stdlib reals.

### PR DESCRIPTION
**Pull request by VincentSe**
Use implicit arguments for ConstructiveReals. Move ConstructiveReals into new directory Abstract. Remove imports of implementations inside those Abstract files.

**Kind:** feature.

### CODE REVIEW

**Comment by Zimmi48** [comment]
Note that, as this is changing code that was already released, all incompatibilities should be clearly documented in a changelog entry.

**Comment by VincentSe** [comment]
@Zimmi48 Changelog added.

**Comment by VincentSe** [comment]
Does anyone understand the build error in VST ?
```
The term "tc_expr Delta a rho" has type "mpred"
while it is expected to have type "bool".
```
Is VST currently broken ? Shall I restart the continuous integration ?

VST does use stdlib real numbers, but I don't think it uses the new constructive reals. My last commit only moves 4 files related to constructive Cauchy reals into their own new directory.

**Comment by MSoegtropIMC** [comment]
> Does anyone understand the build error in VST ?

I can look into this. In general VST has a lot of automation and it improves over time what it can do automatically and what it leaves to the user. This tends to be a bit fragile.

Here at first glance it looks like it automatically simplified a separation logic predicate (Type mpred) to a bool but the proof script still tries to solve the mpred.

**Comment by VincentSe** [comment]
@MSoegtropIMC Thanks. That same VST error happened on many different recent pull requests, it's probably unrelated to this particular PR.

**Comment by MSoegtropIMC** [comment]
> That same VST error happened on many different recent pull requests, it's probably unrelated to this particular PR.

Yes. I think it is not a good idea to use master for VST - cause of the automation it is a bit of a special case and I don't think they have mandatory CI on their master branch. I will make a PR to pin it to a suitable version.

**Review by herbelin — COMMENTED** [review]
@VincentSe:

This is a too big PR for me to review in details, and which provides much more than what the title says (in particular, it contains new contents from #9185 as far as I understood), so could you summarize into more details what it does, in particular with respect to the following points:
- new files/results wrt the existing; for instance, I see results on absolute value, limits, sums; what is new, what is a constructive approach of the existing (e.g. what is the difference between `ConstructiveSum.sum_f_R0` and `Rfunctions.sum_f_R0`, and why twice the same name);
- rôle of `Abstract` directory: this is apparently the signature of constructive reals (unique up to isomorphism);
- rôle of `Cauchy` directory: this is apparently the implementation using Cauchy sequences.

Could you also add a bit more comments at the top of the files (and even inside files). For instance, `ConstructiveAbs.v` has the only short comment `(* Absolute value *)`. Looking at it, I understood that it provides properties of the absolute value based only on the abstract signature but that should be said explicitly, using `coqdoc` parsable comments. Also, would it make sense to rename the file into something like `ConstructiveAbsProperties.v` to emphasize that it is not about an axiomatization of absolute value but about results (see how files are named in `FSets` for instance). Then, probably, other files will deserve the suffix `Properties` too. Note that I'm not saying that this is necessarily the exact way to go: what I'm saying is that this would help browsing the library to know what a file is about from the name (which is not always the case currently for `Reals`, I agree). For instance, it was not clear to me at first glance that `Abstract/ConstructiveAbs.v` was dependent on `Abstract/ConstructiveReals.v` and not the reverse (since we generally have the convention that the file with the less specific name is the collection of all the files of the directory).

Also, the `Reals` library starting to be large, we may think to make the `index-list.html.template` more informative, with sections, subsections, ... Don't hesitate if you have a proposal to make in this direction. Or, otherwise, this can be discussed in group, e.g. with @zimmi48 and @jfehrle who are reworking the documentation.

Another question is how much can this development be connected to other directions under consideration?

For instance, there is the question of a flat structure for reals vs a structure hightlighting the underlying substructures: field, ring, ordered type, ... 

My personal feeling is to prefer a flat structure (*): reals are enough pervasive in the mathematical/physical/computing world to justify that we define them in a stand-alone way. On the other side, we have to be careful that generic lemmas about fields, rings, groups, will apply to the reals, so I suspect that at some time, canonical equipments of `R` as a field, as a ring, etc. will have to provided when such a library of algebraic hierarchy will (hopefully) be available. I'm not particularly an expert on using an algebraic hierarchy, but I guess this will not pose a problem though.

(*) even if not so flat here, since it reuses the definition of `ring_theory`.

Otherwise, this is a very impressive work, and even more when looking also at #9185. I personally consider that it would be great to have all these results available as part of a large library on reals, maybe one day in relation with CoRN, Coquelicot, the MILC's contribution, ...

**Comment by Zimmi48** [comment]
> Also, the `Reals` library starting to be large, we may think to make the `index-list.html.template` more informative, with sections, subsections, ... Don't hesitate if you have a proposal to make in this direction. Or, otherwise, this can be discussed in group, e.g. with @Zimmi48 and @jfehrle who are reworking the documentation.

The documentation of the stdlib is out of our scope of concern for now. I mean that's definitely something that we will have to look into at some point, but we are still too far away from any concrete proposal to improve it, and in the meantime, there is much to do on the reference manual alone.

**Comment by VincentSe** [comment]
Thanks @herbelin for taking the time to read this, even rapidly.

Apart from sorting the files into subdirectories, this PR also reverses the isomorphic approach that I previously took. In file `ConstructiveRealsMorphisms.v`, I prove that all constructive reals implementations are isomorphic, which allowed me to transport results proved on Cauchy reals to any other implementation.

While this is correct, it makes the imports somewhat complicated : Cauchy reals were leaked everywhere. Instead folder `Abstract` is now completely stand-alone, it does not import Cauchy reals anymore. Thus I had to re-prove certain simple lemmas abstractly, concerning absolute value, minimum, maximum, sums and limits. I'll rename `ConstructiveSum.CRsum` instead of `sum_f_R0`. The latter is in the classical part of the stdlib, we have to keep it for backward compatibility. 

I read library `FSets`, and indeed I have an interface in file `ConstructiveReals.v`, all other files being properties. But that is a bit boring to see all files in a directory named "Properies" except just one file, don't you think ? Besides interfaces in `FSets` are module types, while `ConstructiveReals` is a record. I never really understood the difference between those 2 patterns.

Concerning the algebraic hierarchy, I agree with you. Real numbers are so fundamental that they should be defined in a stand-alone way, with all fields in the same file, so that we do not have to browse a dozen files in a library to reconstruct the total picture. This has the other advantage that abstract reals depend on nothing, they can be moved into any other library if we decide to split the stdlib in the future. The ring fields CRisRing and CRisRingExt are Cauchy reals legacy, I can split them into stand-alone fields if you prefer.

In mathematics, properties are often separated from definitions, as you suggest with the "Properties" renaming of the files. Implementing structures (algebraic or not) is properties, these should be proved after the definitions.

The ring tactic would be appreciated for abstract reals, but it is not possible at the moment. We need a parametric `forall (R : ConstructiveReals), Add Ring R`.

In my opinion, the large libraries on real numbers already exist. It is Coquelicot for the classical part and CoRN for the constructive part. I will try to connect them, by allowing Coquelicot to import lemmas from CoRN, so that classical theorems merely extend constructive proofs by calling a few excluded middles. I do not understand why the new mathcomp/analysis is currently being developed.

**Comment by Zimmi48** [comment]
>I do not understand why the new mathcomp/analysis is currently being developed.

We can ask @CohenCyril.

BTW, @spitters said [on Gitter](https://gitter.im/coq/coq?at=5e713e9b63cbdf33ea407d45):

>I understand [Coquelicot and math-comp analysis] are planned to converge

**Comment by VincentSe** [comment]
Hi @herbelin, each file now starts with a coqdoc comment and sum_f_R0 is renamed. Is there something else that blocks merging ?

File: doc/changelog/10-standard-library/11725-cleanup-reals.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+- **Changed:**
+  Use implicit arguments for ConstructiveReals. Move ConstructiveReals
+  into new directory Abstract. Remove imports of implementations inside
+  those Abstract files. Move Cauchy reals in new directory Cauchy.
```
Comment (herbelin) [inline doc/changelog/10-standard-library/11725-cleanup-reals.rst]:
```suggestion
  those ``Abstract`` files. Move implementation by means of Cauchy sequences in new directory ``Cauchy``.
```

File: doc/changelog/10-standard-library/11725-cleanup-reals.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+- **Changed:**
+  Use implicit arguments for ConstructiveReals. Move ConstructiveReals
```
Comment (herbelin) [inline doc/changelog/10-standard-library/11725-cleanup-reals.rst]:
```suggestion
Use implicit arguments for ``ConstructiveReals``. Move ``ConstructiveReals``
```

By "Use implicit arguments", you mean that the `R:ConstructiveReals` argument of lemmas is now implicit, right. Maybe, this can be told more explicitly.

File: doc/changelog/10-standard-library/11725-cleanup-reals.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  into new directory Abstract. Remove imports of implementations inside
```
Comment (herbelin) [inline doc/changelog/10-standard-library/11725-cleanup-reals.rst]:
```suggestion
  into new directory ``Abstract``. Remove imports of implementations inside
```

**Comment by herbelin** [comment]
> Besides interfaces in FSets are module types, while ConstructiveReals is a record. I never really understood the difference between those 2 patterns.

At some point of the history of modules, there was an hope to use module types to represent mathematical structures. I don't think there are people still going this way. My own view is that functors are more about library management: they are ok for deriving concrete instances, together with a whole environment of notations, tactics, etc., as `FSets` allows, but they are not there to give generic theorems (no way to turn a quantification over a module type into a `forall`).

> The ring tactic would be appreciated for abstract reals, but it is not possible at the moment. We need a parametric `forall (R : ConstructiveReals), Add Ring R`.

Interesting remark! Actually, I suspect it is possible to bypass `Add Ring` by calling instead the Ltac function `Ring_gen`. Indeed, `Add Ring` is just registrering a structure in a global table with `ring` looking this structure up based on the type of the expression to simply. Then, `ring` calls the Ltac function `Ring_gen` with this structure. So, it should be possible to pass a `ring_theory` directly to `Ring_gen` without going via the global table. Some wrapper would need to be written to turn a `ring_theory` into the format expected by `Ring_gen` though (apparently a Church-style encoding of tuples). This is in `theories/setoid_ring/Ring_tac.v`.

> But that is a bit boring to see all files in a directory named "Properties" except just one file, don't you think ?

Somehow yes. I don't have a good solution to propose, so let's keep as it is.

**Review by herbelin — APPROVED** [review]
> Is there something else that blocks merging ?

I made a few comments on the change log, which I prefer are addressed. I'll make a few other remarks and suggestions on the form, but nothing blocking per se.

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 4]
```diff
@@ -0,0 +1,1150 @@
...
+Notation "x <= y" := (CRle _ x y) : ConstructiveReals.
+Notation "x <= y <= z" := (CRle _ x y /\ CRle _ y z) : ConstructiveReals.
+Notation "x < y < z"   := (prod (CRlt _ x y) (CRlt _ y z)) : ConstructiveReals.
+Notation "x == y" := (CReq _ x y) : ConstructiveReals.
+Notation "0" := (CRzero _) : ConstructiveReals.
+Notation "1" := (CRone _) : ConstructiveReals.
+Notation "x + y" := (CRplus _ x y) : ConstructiveReals.
+Notation "- x" := (CRopp _ x) : ConstructiveReals.
+Notation "x - y" := (CRminus _ x y) : ConstructiveReals.
+Notation "x * y" := (CRmult _ x y) : ConstructiveReals.
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: An alternative is to use the `where` clause of `Record` so that the notations can be used at once in the definition of the record.

Also, that'd be consistent in style to come up with a notation for `CRinv`, maybe something like `x /_[ H ] y`, or `x / y [ H ]`, or `x / y by H`. Or maybe simply using:
```
Notation "x / y" :=  (CRinv x y ltac:(auto)).
```

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 5]
```diff
@@ -0,0 +1,1150 @@
...
+
+
+Require Import QArith Qabs Qround.
+
+Definition isLinearOrder {X : Set} (Xlt : X -> X -> Set) : Set
+  := (forall x y:X, Xlt x y -> Xlt y x -> False)
+     * (forall x y z : X, Xlt x y -> Xlt y z -> Xlt x z)
+     * (forall x y z : X, Xlt x z -> Xlt x y + Xlt y z).
+
+Record ConstructiveReals : Type :=
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: It is more or less the usage to use `Structure` rather than `Record` when dealing with algebraic structure. They are synonyms, so, it is a bit a question of taste.

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 6]
```diff
@@ -0,0 +1,1150 @@
...
+
+Lemma CRle_lt_trans : forall {R : ConstructiveReals} (r1 r2 r3 : CRcarrier R),
+    r1 <= r2 -> r2 < r3 -> r1 < r3.
+Proof.
+  intros. destruct (CRltLinear R).
+  destruct (s r2 r1 r3 H0). contradiction. apply c.
+Qed.
+
+Lemma CRlt_le_trans : forall {R : ConstructiveReals} (r1 r2 r3 : CRcarrier R),
+    r1 < r2 -> r2 <= r3 -> CRlt R r1 r3.
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: Is there any reason to use `CRlt` and not `<`?

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 7]
```diff
@@ -0,0 +1,1150 @@
...
+
+Lemma CRlt_trans : forall {R : ConstructiveReals} (x y z : CRcarrier R),
+    x < y -> y < z -> x < z.
+Proof.
+  intros. apply (CRlt_le_trans _ y _ H).
+  apply CRlt_asym. exact H0.
+Defined.
+
+Lemma CRlt_trans_flip : forall {R : ConstructiveReals} (x y z : CRcarrier R),
+    CRlt R y z -> CRlt R x y -> CRlt R x z.
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: Idem.

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 8]
```diff
@@ -0,0 +1,1150 @@
...
+Qed.
+
+Lemma CRplus_opp_r : forall {R : ConstructiveReals} (x : CRcarrier R),
+    x + - x == 0.
+Proof.
+  intros. destruct (CRisRing R). apply Ropp_def.
+Qed.
+
+Lemma CRopp_0 : forall {R : ConstructiveReals},
+    CReq R (- 0) 0.
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: Idem, is there any reason to use `CReq` rather than `==`?

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 9]
```diff
@@ -0,0 +1,1150 @@
...
+Proof.
+  intros.
+  destruct (CRisRingExt R). clear Rmul_ext Ropp_ext.
+  pose proof (Radd_ext
+                (CRopp R r) (CRopp R r) (CReq_refl _)
+                _ _ H).
+  destruct (CRisRing R).
+  apply (CReq_trans r1) in H0.
+  apply (CReq_trans _ _ _ H0).
+  apply (CReq_trans _ (CRplus R (CRplus R (CRopp R r) r) r2)).
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: Not sure it would make a very big difference, but `transitivity r1` or `stepl`/`stepr` are idioms which have been proposed for this kind of stepwise reasoning. They are widely used though, so using them may also look unconventional (this is more a question for stdlib2...)

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 10]
```diff
@@ -0,0 +1,1150 @@
...
+    0 < r
+    -> r1 <= r2
+    -> r1 * r <= r2 * r.
+Proof.
+  intros. intro abs. apply CRmult_lt_reg_r in abs.
+  contradiction. exact H.
+Qed.
+
+Lemma CRmult_eq_reg_r : forall {R : ConstructiveReals} (r r1 r2 : CRcarrier R),
+    CRapart R 0 r
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: CoRN had a notation for `CRapart` iirc? What's your opinion about abbreviating it?

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 11]
```diff
@@ -0,0 +1,1150 @@
...
+
+Lemma CRlt_minus : forall {R : ConstructiveReals} (x y : CRcarrier R),
+    x < y -> 0 < y - x.
+Proof.
+  intros. rewrite <- (CRplus_opp_r x).
+  apply CRplus_lt_compat_r. exact H.
+Qed.
+
+Lemma CR_of_Q_le : forall {R : ConstructiveReals} (r q : Q),
+    Qle r q
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: None of them is optimal but why `Qle r q` rather than `(r <= q)%Q`?

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 12]
```diff
@@ -0,0 +1,1150 @@
...
+
+Add Parametric Morphism {R : ConstructiveReals} : (CR_of_Q R)
+    with signature Qeq ==> CReq R
+      as CR_of_Q_morph.
+Proof.
+  split; apply CR_of_Q_le; rewrite H; apply Qle_refl.
+Qed.
+
+Lemma eq_inject_Q : forall {R : ConstructiveReals} (q r : Q),
+    CR_of_Q R q == CR_of_Q R r -> Qeq q r.
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: Idem for `Qeq`.

File: theories/Reals/Abstract/ConstructiveReals.v [hunk 13]
```diff
@@ -0,0 +1,1150 @@
...
+  apply (CReq_trans _ (CRzero R)).
+  apply (CReq_trans _ (CR_of_Q R (q-q))).
+  apply CReq_sym, CR_of_Q_plus.
+  apply (CReq_trans _ (CR_of_Q R 0)).
+  apply CR_of_Q_morph. ring. apply CR_of_Q_zero.
+  apply CReq_sym. apply (CRisRing R).
+Qed.
+
+Lemma CR_of_Q_pos : forall {R : ConstructiveReals} (q:Q),
+    Qlt 0 q -> CRlt R (CRzero R) (CR_of_Q R q).
```
Comment (herbelin) [inline theories/Reals/Abstract/ConstructiveReals.v]: Idem for `Qlt`. And why not `<` and `0` on `R` (is it because it would be too ambiguous)?

**Review by herbelin — CHANGES_REQUESTED** [review]
@VincentSe: I then made extra comments on the form regarding `ConstructiveReals`.

Most comments are to give perspectives but I think that it would be good already in this PR to have a uniform use of the `<`, `<=`, `==` notations in the statements (thus the "requested changes").

Another question (by curiosity) is also why to make `R` implicit and not more arguments. For instance, in `CRlt_asym`, `x` and `y` could somehow be made implicit too.

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 10 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: That notation is a problem. Usually it is `x # y`, but in the stdlib that clashes with the rational `p # q`, meaning `p / q`. What do you think, we use `#` or not ?

Comment (herbelin) [reply theories/Reals/Abstract/ConstructiveReals.v]: `#` is apparently Brouwer's notation and is quite standard. I see that Bridges is using something like `|><|` (`▷◁` ??). We could also try to change `#` from `QArith`  (whose meaning is not quite obvious) but I'm afraid of incompatibilies. 

It could be also `><`? Or `##`? Or `≷`?

Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: I like `><` which literally says what it means : `x < y \/ y < x`

Comment (SkySkimmer) [reply theories/Reals/Abstract/ConstructiveReals.v]: HoTTClasses uses `≶` btw
https://github.com/HoTT/HoTT/blob/705244361395ca44de10554691ad80975cf60df6/theories/Classes/interfaces/canonical_names.v#L39

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 4 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: Can you show me an example of where clause here ? The documentation is not very clear
https://coq.inria.fr/refman/language/gallina-extensions.html

Comment (herbelin) [reply theories/Reals/Abstract/ConstructiveReals.v]: Here is an example:
```
Record t :=
  { op : nat -> nat -> nat where "x + y" := (op x y);
    comm : forall x y, x + y = y + x }.
```
In passing, the online doc seems a bit outdated but it is correct in master (afaics): the `where` clause can also be applied to defined fields.

Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: Now Structure ConstructiveReals compiles with its notations in where clauses, but those no longer take precedence over Q's notations. How do we put notations in scope ConstructiveReals, which is declared after Structure ConstructiveReals ?

Comment (herbelin) [reply theories/Reals/Abstract/ConstructiveReals.v]: Put the `Declare Scope ConstructiveReals.` before the `Record` and add `: ConstructiveReals` at the end of the `where` clause.

Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: Nope, still does not work. I did this
```
    CRlt : CRcarrier -> CRcarrier -> Set
    where "x < y" := (CRlt x y) : ConstructiveReals;
```
and then Coq complains on `x < y` in `CRlt_asym` a few lines below. Is it a bug in notation scopes ?

Comment (herbelin) [reply theories/Reals/Abstract/ConstructiveReals.v]: Oh, I see. We need the `Local Open Scope ConstructiveReals.` also to be done before the `Record`.

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 5 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: done

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 6 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: done

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 7 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: done

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 8 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: done

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 9 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: done

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 10 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: @SkySkimmer Thanks, I used ≶ also

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 13 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: done

**Comment by VincentSe** [comment]
@herbelin I did all the improvements you suggested, expect the `where` clause in `Structure ConstructiveReals`, which conflicts with Q's notations on the several different versions I tried.

File: theories/Reals/Abstract/ConstructiveReals.v (hunk 4 above)
Comment (VincentSe) [reply theories/Reals/Abstract/ConstructiveReals.v]: No. Adding this still conflicts with Q's notations
```
Declare Scope ConstructiveReals.

Local Open Scope ConstructiveReals.

Structure ConstructiveReals : Type :=
```

**Comment by herbelin** [comment]
> I did all the improvements you suggested, expect the `where` clause in `Structure ConstructiveReals`, which conflicts with Q's notations on the several different versions I tried.

Indeed. I thought that the notations defined using the `where` clause were surviving the definition of the `Structure`, but it does not. I'll see if I can do something about it.

Thanks for all the changes. I plan to merge by Monday, if anyone wants to say something by then.

PS: there seems to be a compilation problem with the change log. Indentation is meaningful and I believe it requires 2 spaces in the second line.

**Review by herbelin — APPROVED** [review]
